### PR TITLE
Revoke refresh token

### PR DESCRIPTION
**Pull request by avishalom**
Adding the revokeRefreshToken, 
and the ability to check revoked when validating a token.

### CODE REVIEW

File: firebase_admin/auth.py [hunk 1]
```diff
@@ -85,17 +85,27 @@ def verify_id_token(id_token, app=None):
...
-    return token_generator.verify_id_token(id_token)
-
+    verified_claims = token_generator.verify_id_token(id_token)
+    if check_revoked:
+        user = get_user(verified_claims.get('uid'), app)
+        if  verified_claims.get('iat') < user.tokens_valid_after_time:
+            raise AuthError('ID_TOKEN_REVOKED', 'The Firebase ID token has been revoked.')
+    return verified_claims
+
+def revoke_refresh_tokens(uid, app=None):
```
Comment (hiranya911) [inline firebase_admin/auth.py]: Add docstring. Also update the docstring of `update_user()` with the new kwarg

File: firebase_admin/auth.py [hunk 2]
```diff
@@ -430,6 +440,16 @@ def disabled(self):
         """
         return bool(self._data.get('disabled'))
 
+    @property
+    def tokens_valid_after_time(self):
+        """Returns the time before which tokens are invalid.
+
+        Returns:
+            int: Timestap in seconds since the epoch. All tokens issued before that time
```
Comment (hiranya911) [inline firebase_admin/auth.py]: Typo: Timestap

File: tests/test_auth.py [hunk 3]
```diff
@@ -48,6 +48,9 @@
 
 MOCK_GET_USER_RESPONSE = testutils.resource('get_user.json')
 MOCK_LIST_USERS_RESPONSE = testutils.resource('list_users.json')
+_REVOKED_TOKENS_USER = json.loads(MOCK_GET_USER_RESPONSE)
+_REVOKED_TOKENS_USER['users'][0]['validSince'] = str(int(time.time())+1)
+MOCK_GET_USER_REVOKED_TOKENS_RESPONSE = json.dumps(_REVOKED_TOKENS_USER)
```
Comment (hiranya911) [inline tests/test_auth.py]: Perhaps introduce a helper function here:

```
MOCK_GET_USER_REVOKED_TOKENS_RESPONSE = _revoked_tokens_response()
```

File: firebase_admin/auth.py [hunk 4]
```diff
@@ -85,17 +85,27 @@ def verify_id_token(id_token, app=None):
...
+    verified_claims = token_generator.verify_id_token(id_token)
+    if check_revoked:
+        user = get_user(verified_claims.get('uid'), app)
+        if  verified_claims.get('iat') < user.tokens_valid_after_time:
+            raise AuthError('ID_TOKEN_REVOKED', 'The Firebase ID token has been revoked.')
+    return verified_claims
+
+def revoke_refresh_tokens(uid, app=None):
+    user_manager = _get_auth_service(app).user_manager
+    user_manager.update_user(uid, tokens_valid_after_time=str(int(time.time())))
```
Comment (hiranya911) [inline firebase_admin/auth.py]: Unnecessary wrapping in a str IMO

File: firebase_admin/auth.py [hunk 5]
```diff
@@ -430,6 +440,16 @@ def disabled(self):
...
 
+    @property
+    def tokens_valid_after_time(self):
+        """Returns the time before which tokens are invalid.
+
+        Returns:
+            int: Timestap in seconds since the epoch. All tokens issued before that time
+                 are considered revoked.
+        """
+        return int(self._data.get('validSince', 0))
```
Comment (hiranya911) [inline firebase_admin/auth.py]: We seem to return 0 by default.

1. Is that consistent with the behavior in Node SDK (perhaps `None` is more appropriate here)?
2. We should document this in the docstring.

File: tests/test_auth.py [hunk 6]
```diff
@@ -244,8 +244,47 @@ def test_valid_token(self, authtest, id_token):
         assert claims['admin'] is True
         assert claims['uid'] == claims['sub']
 
-    @pytest.mark.parametrize('id_token', invalid_tokens.values(),
-                             ids=list(invalid_tokens))
+    @pytest.mark.parametrize('id_token', valid_tokens.values(), ids=list(valid_tokens))
+    def test_valid_token_check_revoked(self, user_mgt_app_for_verify_token, id_token):
+        _instrument_user_manager(user_mgt_app_for_verify_token, 200, MOCK_GET_USER_RESPONSE)
```
Comment (hiranya911) [inline tests/test_auth.py]: Also need a test case for when the `UserRecord` contains a future timestamp.

File: tests/test_auth.py [hunk 7]
```diff
@@ -244,8 +244,47 @@ def test_valid_token(self, authtest, id_token):
...
+    @pytest.mark.parametrize('id_token', valid_tokens.values(), ids=list(valid_tokens))
+    def test_revoked_token_check_revoked(self, user_mgt_app_for_verify_token, id_token):
+        _instrument_user_manager(user_mgt_app_for_verify_token, 200,
+                                 MOCK_GET_USER_REVOKED_TOKENS_RESPONSE)
+
+        with pytest.raises(auth.AuthError) as excinfo:
+            auth.verify_id_token(id_token, app=user_mgt_app_for_verify_token,
+                                 check_revoked=True)
+
+        assert excinfo.value.code == 'ID_TOKEN_REVOKED'
```
Comment (hiranya911) [inline tests/test_auth.py]: Check against the constant once you have it

File: tests/test_auth.py [hunk 8]
```diff
@@ -244,8 +244,47 @@ def test_valid_token(self, authtest, id_token):
...
+    @pytest.mark.parametrize('id_token', valid_tokens.values(), ids=list(valid_tokens))
+    def test_revoked_token_do_not_check_revoked(self, user_mgt_app_for_verify_token, id_token):
+        _instrument_user_manager(user_mgt_app_for_verify_token, 200,
+                                 MOCK_GET_USER_REVOKED_TOKENS_RESPONSE)
+        claims = auth.verify_id_token(id_token, app=user_mgt_app_for_verify_token,
+                                      check_revoked=False)
+        assert claims['admin'] is True
+        assert claims['uid'] == claims['sub']
+
+    def test_revoke_refresh_token(self, user_mgt_app):
```
Comment (hiranya911) [inline tests/test_auth.py]: Add test case for setting this via `update_user()`

File: tests/test_auth.py [hunk 9]
```diff
@@ -287,6 +326,13 @@ def user_mgt_app():
     yield app
     firebase_admin.delete_app(app)
 
+@pytest.fixture(scope='module')
+def user_mgt_app_for_verify_token():
```
Comment (hiranya911) [inline tests/test_auth.py]: Can we just modify the existing `user_mgt_app` fixture to have the `project_id` option?

File: firebase_admin/_user_mgt.py [hunk 10]
```diff
@@ -113,6 +113,20 @@ def validate_photo_url(cls, photo_url):
         except Exception:
             raise ValueError('Malformed photo URL: "{0}".'.format(photo_url))
 
+    @classmethod
+    def validate_tokens_valid_after_time(cls, valid_since):
+        if not isinstance(valid_since, six.string_types) or not valid_since:
```
[the remark this answers is not in the document]
Comment (avishalom) [reply firebase_admin/_user_mgt.py]: the other timestamps in the UserMetadata are also returned as quoted strings. (passwordUpdate is a float in the response, but all other timestamps are quoted.)

File: firebase_admin/auth.py (hunk 4 above)
Comment (avishalom) [reply firebase_admin/auth.py]: It works without it, but this conforms to the other timestamps (the golang json tags also wrap this field in quotes.)

File: CHANGELOG.md [hunk 11]
```diff
@@ -1,11 +1,19 @@
 # Unreleased
 - 
+### Token revokaction
+- [added] The ['verify_id_token(...)'](https://firebase.google.com/docs/reference/admin/python/firebase_admin.auth#verify_id_token)
+  method now accepts an optional `check_revoked` parameter, when `True`, an
```
Comment (hiranya911) [inline CHANGELOG.md]: parameter. When `True`...

File: CHANGELOG.md [hunk 12]
```diff
@@ -1,11 +1,19 @@
 # Unreleased
 - 
+### Token revokaction
+- [added] The ['verify_id_token(...)'](https://firebase.google.com/docs/reference/admin/python/firebase_admin.auth#verify_id_token)
+  method now accepts an optional `check_revoked` parameter, when `True`, an 
+  additional check is performed to see if the token has been revoked. 
+- [added] A new method ['auth.revoke_refresh_tokens(uid)'](https://firebase.google.com/docs/reference/admin/python/firebase_admin.auth#revoke_refresh_tokens)
+  has been added to invalidate all tokens issued before the current second.
```
Comment (hiranya911) [inline CHANGELOG.md]: ...issued to a user.

File: CHANGELOG.md [hunk 13]
```diff
@@ -1,11 +1,19 @@
...
+- [added] A new property `tokens_valid_after_time` has been added to the ['UserRecord'](https://firebase.google.com/docs/reference/admin/python/firebase_admin.auth#userrecord),
+  it denotes the time in epoch milliseconds before which tokens are not valid.
```
Comment (hiranya911) [inline CHANGELOG.md]: which denotes..

File: firebase_admin/_user_mgt.py [hunk 14]
```diff
@@ -113,6 +113,20 @@ def validate_photo_url(cls, photo_url):
         except Exception:
             raise ValueError('Malformed photo URL: "{0}".'.format(photo_url))
 
+    @classmethod
+    def validate_valid_since(cls, valid_since):
+        if not isinstance(valid_since, six.string_types) or not valid_since:
```
Comment (hiranya911) [inline firebase_admin/_user_mgt.py]: Why is this a string? It should just accept a timestamp.

File: firebase_admin/auth.py [hunk 15]
```diff
@@ -85,17 +86,27 @@ def verify_id_token(id_token, app=None):
     Args:
       id_token: A string of the encoded JWT.
       app: An App instance (optional).
+      check_revoked: A boolean indicating whether the token has been revoked (optional).
```
Comment (hiranya911) [inline firebase_admin/auth.py]: whether to check if the token has been...

File: firebase_admin/auth.py [hunk 16]
```diff
@@ -85,17 +86,27 @@ def verify_id_token(id_token, app=None):
...
+    verified_claims = token_generator.verify_id_token(id_token)
+    if check_revoked:
+        user = get_user(verified_claims.get('uid'), app)
+        if  verified_claims.get('iat') * 1000 < user.tokens_valid_after_time:
+            raise AuthError(_ID_TOKEN_REVOKED, 'The Firebase ID token has been revoked.')
+    return verified_claims
+
+def revoke_refresh_tokens(uid, app=None):
+    user_manager = _get_auth_service(app).user_manager
+    user_manager.update_user(uid, valid_since=str(int(time.time())))
```
Comment (hiranya911) [inline firebase_admin/auth.py]: Why wrap this in str()? Just pass the int timestamp.

File: firebase_admin/auth.py [hunk 17]
```diff
@@ -476,12 +497,16 @@ def __init__(self, data):
 
     @property
     def creation_timestamp(self):
+        """ Creation timestamp in epoch milliseconds.
+        """
```
Comment (hiranya911) [inline firebase_admin/auth.py]: Add the `Returns:` section too while we are at it?

File: CHANGELOG.md [hunk 18]
```diff
@@ -2,11 +2,12 @@
...
-  additional check is performed to see if the token has been revoked. 
+  method now accepts an optional `check_revoked` parameter. When `True`, an 
+  additional check is performed to see whether the token has been revoked. 
 - [added] A new method ['auth.revoke_refresh_tokens(uid)'](https://firebase.google.com/docs/reference/admin/python/firebase_admin.auth#revoke_refresh_tokens)
-  has been added to invalidate all tokens issued before the current second.
-- [added] A new property `tokens_valid_after_time` has been added to the ['UserRecord'](https://firebase.google.com/docs/reference/admin/python/firebase_admin.auth#userrecord),
+  has been added to invalidate all tokens issued to a user before the current second.
+- [added] A new property `tokens_valid_after_timestamp` has been added to the   
+['UserRecord'](https://firebase.google.com/docs/reference/admin/python/firebase_admin.auth#userrecord),
   it denotes the time in epoch milliseconds before which tokens are not valid.
```
Comment (hiranya911) [inline CHANGELOG.md]: which denotes...

File: firebase_admin/_user_mgt.py [hunk 19]
```diff
@@ -115,17 +115,12 @@ def validate_photo_url(cls, photo_url):
...
-        try:
-            int(valid_since)
-        except ValueError:
-            raise ValueError('Malformed Valid Since: "{0}" '
-                             'must be an interger.'.format(valid_since))
-        except Exception:
-            raise ValueError('Malformed Valid Since "{0}".'.format(valid_since))
+                'Invalid time string for: "{0}". Valid Since must be an int'.format(valid_since))
+        if int(valid_since) <= 0:
+            raise ValueError('Valid Since: must be a positive interger. {0}'.format(valid_since))
```
Comment (hiranya911) [inline firebase_admin/_user_mgt.py]: `Invalid valid_since: ...` ?

File: firebase_admin/auth.py [hunk 20]
```diff
@@ -100,13 +100,24 @@ def verify_id_token(id_token, app=None, check_revoked=False):
...
+    """Revokes all refresh tokens for an existing user.
+
+    revoke_refresh_tokens updates the user's tokens_valid_after_timestamp to the current UTC
+    in seconds since the epoch. It is important that the server on which this is called has its
+    clock set correctly and synchronized.
+
+    While this revokes all sessions for a specified user and disables any new ID tokens for
+    existing sessions from getting minted, existing ID tokens may remain active until their
+    natural expiration (one hour). To verify that ID tokens are revoked, use
+    `verify_id_token(idToken, check_revoked=true)`.
```
Comment (hiranya911) [inline firebase_admin/auth.py]: `check_revoked=True`

File: tests/data/list_users.json [hunk 21]
```diff
@@ -49,7 +49,7 @@
         "passwordUpdatedAt" : 1.494364393E+12,
         "validSince" : "1494364393",
         "disabled" : false,
-        "createdAt" : "1234567890",
+        "createdAt" : "1234567890000",
         "customAttributes" : "{\"admin\": true, \"package\": \"gold\"}"
     } ]
 }
```
Comment (hiranya911) [inline tests/data/list_users.json]: Not really related to your change, but lets add a new line here.

**Pull request merged**